substrate/client/network/sync/src/engine.rs

```diff
@@ -783,120 +806,166 @@ where
-			self.network_service.report_peer(who, rep::GENESIS_MISMATCH);
+		self.chain_sync.peer_disconnected(&peer);
+		self.event_streams
+			.retain(|stream| stream.unbounded_send(SyncEvent::PeerDisconnected(peer)).is_ok());
+	}
 
-			if self.important_peers.contains(&who) {
+	/// Validate received handshake.
+	fn validate_handshake(
+		&mut self,
+		peer_id: &PeerId,
+		handshake: Vec<u8>,
+	) -> Result<BlockAnnouncesHandshake<B>, bool> {
+		log::trace!(target: LOG_TARGET, "Validate handshake for {peer_id}");
+
+		let handshake = <BlockAnnouncesHandshake<B> as DecodeAll>::decode_all(&mut &handshake[..])
+			.map_err(|error| {
+				log::debug!(target: LOG_TARGET, "Failed to decode handshake for {peer_id}: {error:?}");
+				false
+			})?;
+
+		if handshake.genesis_hash != self.genesis_hash {
+			if self.important_peers.contains(&peer_id) {
 				log::error!(
-					target: "sync",
-					"Reserved peer id `{}` is on a different chain (our genesis: {} theirs: {})",
-					who,
+					target: LOG_TARGET,
+					"Reserved peer id `{peer_id}` is on a different chain (our genesis: {} theirs: {})",
 					self.genesis_hash,
-					status.genesis_hash,
+					handshake.genesis_hash,
 				);
-			} else if self.boot_node_ids.contains(&who) {
+			} else if self.boot_node_ids.contains(&peer_id) {
 				log::error!(
-					target: "sync",
-					"Bootnode with peer id `{}` is on a different chain (our genesis: {} theirs: {})",
-					who,
+					target: LOG_TARGET,
+					"Bootnode with peer id `{peer_id}` is on a different chain (our genesis: {} theirs: {})",
 					self.genesis_hash,
-					status.genesis_hash,
+					handshake.genesis_hash,
 				);
 			} else {
 				log::debug!(
-					target: "sync",
+					target: LOG_TARGET,
 					"Peer is on different chain (our genesis: {} theirs: {})",
-					self.genesis_hash, status.genesis_hash
+					self.genesis_hash,
+					handshake.genesis_hash
 				);
 			}
 
-			return Err(())
+			return Err(true)
 		}
 
-		let no_slot_peer = self.default_peers_set_no_slot_peers.contains(&who);
-		let this_peer_reserved_slot: usize = if no_slot_peer { 1 } else { 0 };
+		Ok(handshake)
+	}
 
-		// make sure to accept no more than `--in-peers` many full nodes
-		if !no_slot_peer &&
-			status.roles.is_full() &&
-			inbound && self.num_in_peers == self.max_in_peers
-		{
-			log::debug!(target: "sync", "All inbound slots have been consumed, rejecting {who}");
-			return Err(())
+	/// Validate connection.
+	// NOTE Returning `Err(bool)` is a really ugly hack to work around the issue
+	// that `ProtocolController` thinks the peer is connected when in fact it can
+	// still be under validation. If the peer has different genesis than the
+	// local node the validation fails but the peer cannot be reported in
+	// `validate_connection()` as that is also called by
+	// `ValiateInboundSubstream` which means that the peer is still being
+	// validated and banning the peer when handling that event would
+	// result in peer getting dropped twice.
```

Comment:
Why would we receive `NotificationStreamOpened` for a peer we rejected in `ValidateInboundSubstream`? I probably don't understand why `validate_connection()` can be called twice for a peer we rejected during the substream validation.

Reply:
This function is shared by both `ValidateInboundSubstream` and `NotificationStreamOpen`. We only validate inbound substreams so if we initiated an outbound substream, the resulting inbound substream won't emit `ValidateInboundSubstream` and instead it has to be validated when `NotificationStreamOpen` is received.

Not a good explanation but basically there is asymmetry in the validation even though there is no asymmetry in substreams themselves (we always have both inbound and outbound substream). This probably should be fixed eventually but the situation you're describing cannot (or should not) happen. If the peer is rejected after receiving `ValidateInboundSubstream`, there won't be `NotificationStreamOpened`, unless of course the local node opens a substream to them right after.

Reply:
Thank you, I see now that `ProtocolController` would be notified twice if we both reject the incoming substream and call `disconnect_peer`, so the message of comment is likely about this.